### ISSUE TITLE
[Python] Publish the option --language Python to make it easier to experiment

### ISSUE THREAD

**Issue by Zaid-Ajaj** [fable-compiler/Fable#2577]
### Description

I would love to be able to start using `--language Python` and start experimenting with it. Even though it is not "complete" yet and there are some issues, it will make it easier for users to get a feel of how it works and experiment with what is supported and what isn't. 

I propose to publish it as part of Fable and document as it as _alpha_ by logging the following message when compilation starts 
```
Compiling to Python is still in _alpha_ phase and is currently only for experimentation.
```

I am already getting feature requests to make libraries and tools F#->Python compatible (Femto & Fable.Remoting) so it will be an interesting domain to explore. 

cc @dbrattli 

> P.S. I have seen Fable.Python but I would prefer not having to compile Fable myself if possible 😄 

### Repro code

N/A 

### Expected and actual results

Expected to be able to compile a simple console app
```bash
mkdir FableTest
cd FableTest
dotnet new console -lang F#
dotnet new tool-manifest
dotnet tool install fable
dotnet fable --language Python
python Program.fs.py
```

### Related information

* Fable version: `dotnet fable --version` 3.4.7
* Operating system: Windows 10

**Comment by dbrattli** [comment]
@Zaid-Ajaj Great to hear. There's a lot of things left to fix, but feel that we are approaching critical mass with this port. One gotcha to be aware of is that <OutputType>Exe</OutputType> will currently generate absolute imports in Python, while <OutputType>Library</OutputType> (or default) will generate relative imports. This is because a Python program cannot use relative imports, while a library may not know where it's mounted or imported from, so safer to use relative imports in libraries. I know many do not set output type for e.g net5, but otherwise I would need to detect the presence of an entry-point.

**Comment by Zaid-Ajaj** [comment]
Hi @dbrattli I am having trouble building a library with `fable-py` 

I made the repository [Fable.SimpleJson.Python](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python) which is a modified version of [Fable.SimpleJson](https://github.com/Zaid-Ajaj/Fable.SimpleJson) without the JS bits but I am unable to run the tests. 

```bash
git clone https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python.git
cd ./Fable.SimpleJson.Python
dotnet tool restore
cd ./tests
dotnet fable-py 
python program.py
```
Some problems are as follows:
### Paths are weird
when compiling using `dotnet fable-py` I think it should generate something that works out of the box but here I have to manually edit the imports
```python
# Compiled code: doesn't work
from fable_modulessrc.json_converter import Convert_fromJson
from fable_modulessrc.simple_json import SimpleJson_parseNative
from fable_modulessrc.type_info_converter import create_type_info
```
see how `fable_modules` was concatenated with `src` which doesn't work. 
The only way I tried to make it run was using `--outDir` so that the packages have a common parent package but even then I had to manually edit the imports

### Using `unint32` on a `float`
```fs
// doesn't work
let value = 2.0
let converted = uint32 value
```
compiles to
```python
value = 2
converted = value >>> 0
```
### Union reflection 

The functions `make_union` and `get_union_fields` are not implemented :/ 

@alfonsogarciacaro is it OK to track python related issues here as well?

**Comment by dbrattli** [comment]
@Zaid-Ajaj I'll have a look. The generated code looks to be wrong so I'll try to fix it. But even then, there are a few things you need to be aware of.

Python imports cannot reference out of their package/directory. Thus you cannot import e.g `src` from something in `test` without altering `sys.path`. So project references will usually not work without altering the path, except when using `pytest` for testing. Pytest will itself alter `sys.path` so you find the modules you are looking for. If you don't want to use `pytest` then your test script needs to set `PYTHONPATH` before running `python`.

FYI: tracked by https://github.com/fable-compiler/Fable/pull/2578

**Comment by Zaid-Ajaj** [comment]
> Python imports cannot reference out of their package/directory. Thus you cannot import e.g src from something in test without altering `sys.path`.

Given how common it is in dotnet to reference projects from sibling directories, I think it would be nice if the the compiled project just worked out of the box without having to know about `sys.path`, maybe have the compile the full project such that the entry point package is in the "root" and all other modules are subdirectories:
```
{outDir}
  | 
  | -- src
  | -- tests
  | -- program.py
```
Where `program.py` is the entry point because `dotnet fable-py` was run against the `tests` project.
Note that in this case, the modules will no longer be compiled next to their F# sources. 
What do you think?

**Comment by dbrattli** [comment]
@Zaid-Ajaj  I'm interested in anything that could lower the friction. Have been mostly busy with getting the all the bricks in place for compiling F# to Python, but very interested in input to make it easier to compile into Python. Will look closer into this tomorrow morning. Have btw published new `fable-py` and `Fable.Core.Experimental` so please check if that could bring your project one step further now that Union reflection should work a bit better: https://github.com/fable-compiler/Fable/blob/beyond/tests/Python/TestReflection.fs#L222

**Comment by Zaid-Ajaj** [comment]
Hi @dbrattli I've just updated the compiler and it looks better. There are a bunch of functions missing. I will try to list them here one by one later. I've updated the project now to make it easier to test it: basically removed the tests project and added an entry point to the library at `./src` to avoid path resolution issues for the time being. This is of course not ideal because when I publish the library, I need to remove the entry point 😅 but it will do for now when testing the library

**Comment by dbrattli** [comment]
Hi @Zaid-Ajaj . Yes, I see that's quite a lot of work to get all those functions in place 😬 , but we should eventually get there: https://github.com/fable-compiler/Fable/pull/2580

**Comment by dbrattli** [comment]
@Zaid-Ajaj FYI: Have just published the fable-py.4.0.0-alpha-008.nupkg and Fable.Core.Experimental.4.0.0-alpha-008.nupkg packages, so please try again. This time I think you should hit the JS specific emits. If you can fix, then let's see what's next after that.

**Comment by Zaid-Ajaj** [comment]
@dbrattli Getting _really_ close, I am trying to fix things locally and see if I get the project running. Here are the things that didn't work:

- Missing these functions from `fable_library.reflection.py`
  - `get_generics`
  - `get_function_elements`
  - `get_tuple_elements`
  - `get_enum_underlying_type`
- Missing the `Lazy` type from `fable_library.util.py`
- Cannot use assignment syntax with `Dictionary` i.e. `dict[key] <- value` 
- For some reason, cannot use [this type](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python/blob/master/src/TypeInfo.fs#L21) as key for a dictionary (error unhashable type)
- Files `fable_library.dateoffset.py` and `fable_library.boolean.py` should use relative fable-library imports:
  - `from .types import FSharpRef` instead of `from fable_library.types import FSharpRef` 

I've updated the project again with more fixes and avoided things that I could do without (caching resolved types with dictionaries).

**Comment by dbrattli** [comment]
@Zaid-Ajaj Thanks for the feedback. Might take a few days for me to fix this (work week again) but I'm pretty sure this use-case will bring Fable Python to the next level. Fixes (wip) at https://github.com/fable-compiler/Fable/pull/2581

**Comment by dbrattli** [comment]
@Zaid-Ajaj For the Python port were impl F# Arrays using Python list. Python only have arrays on numeric types. So far using lists worked well, but for `BitConverter` it feels wrong to return a list of bytes when you really want `bytes` or `bytearray` since you are most likely going to send it somewhere. It may also be complicated to use different backing of Arrays depending on the inner type. Any thoughts?

**Comment by dbrattli** [comment]
Did a quick check and native arrays actually seems to work really nice except for e.g `Seq.sort` depends on `Array.sortInPlaceWith` which is not supported for Python arrays. But this should be able to fix by moving the inner part of `sortWith` to e.g Native.fs:

```fs
let sortWith (comparer: 'T -> 'T -> int) (xs: seq<'T>) =
    delay (fun () ->
        let arr = toArray xs
        Array.sortInPlaceWith comparer arr // Note: In JS this sort is stable
        arr |> ofArray
    )
```

**Comment by ncave** [comment]
@dbrattli Pretty much the same as in JS, there are typed numeric arrays and untyped arrays.

**Comment by dbrattli** [comment]
@ncave The problem I now have is that modules such as `Seq.fs` and `List.fs` uses arrays as the backing store for many operations such as sorting, and I will need to use python lists (i.e ResizeArray) instead. Not I can do this without forking Seq.fs, List.fs. But it's probably fine when there is so many changes needed. Hard to make an abstraction that fits any target language. What do you think?

**Comment by ncave** [comment]
@dbrattli Yes, it's difficult to have the same exact F# library code for all languages. I know I have different implementations for Rust for some things (at least for now), as I need to work around missing features. As you can see for JS, some methods that need to make arrays for some purpose take (receive) an array constructor as an additional parameter, so the same code can handle different array types that differ by construction only.

**Comment by dbrattli** [comment]
@Zaid-Ajaj. Need to merge and generate new packages but looks promising!

<img width="632" alt="Screenshot 2021-11-03 at 16 41 04" src="https://user-images.githubusercontent.com/849479/140093531-09121fb6-310b-4fd2-8395-d0883078995f.png">

**Comment by Zaid-Ajaj** [comment]
@dbrattli The test case worked?! This is HUGE 🤯 🤯 🤯 because it means the full reflection cycle has worked: type resolution, parsing, pattern matching and type activation. 

Once the new bits are pushed, I'll start adding a lot more test cases similar to those in Fable.SimpleJson and see if we can iron things out even more

As for publishing the library, I think I will need to apply some gymnastics to the build script: removing the `Program.fs` from the library before pushing the nuget and adding it back once that is done to later test how a nuget package can be consumed (even tho we know Fable.Python works)

**Comment by dbrattli** [comment]
@Zaid-Ajaj Have pushed:

- https://www.nuget.org/packages/fable-py/4.0.0-alpha-009
- https://www.nuget.org/packages/Fable.Core.Experimental/4.0.0-alpha-009

Please try and see if what issues surfaces this time. I'm sure there's still many things to be fixed, but I think it's really fun with the progress here 😄  I have btw not done anything with the `--outDir` issue yet to make sure the default compile have a common parent dir. This last PR btw increased coverage with ~100 additional passing unit-tests

**Comment by Zaid-Ajaj** [comment]
After modifying the implementation to work with Python better I have added a dozen or two of tests and here are the [tests that are failing](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python/blob/master/src/Program.fs#L127-L144) (along with their error message / missing imports)

Furthermore, using `dotnet fable-py` against any project without `<OutputType>Exe</OutputType>` creates invalid relative fable_module imports regardless of `--outDir` which means
 - Using `<OutputType>Library</OutputType>` creates an invalid project
 - Omitting `<OutputType>` altogether creates an invalid project (defaulting to Library?)

To avoid confusion, maybe it would make sense to output all the python sources into one directory such as `./dist` or any given `--outDir` relative to the compiled project and fix the imports from there on?

**Comment by dbrattli** [comment]
@Zaid-Ajaj Thanks for the feedback. First let's discuss absolute vs relative imports. It's a tricky problem that I don't know how to solve in a good way. Did you btw read this? https://github.com/fable-compiler/Fable/issues/2577#issuecomment-955155692

The problem is that we need to use relative imports when building libraries. The library do not know where it's "mounted" e.g inside `fable_modules` when used as a nuget, or outside when using a project reference, or as a Python package (pip). There's also the Fable.Library.fsproj inside fable_library I cannot see how we can generate imports that are not relative, and still be valid. Let's look at an example:

```
mylibrary
     |--- __init__.py
     |--- util.py
     |--- library.py
```

## Problem 1: 

How should e.g `library.py` import  `foo` from  `util.py`?

1. `from util import foo` Not possible. This gives error `ModuleNotFoundError: No module named 'util'`. Only possible if `library.py` was a program (top module).
2. If it's mounted in fable_modules: `from fable_modules.mylibrary.util import foo`
3. If it's a project reference with a common outDir: `from mylibrary.util import foo`
4. Relative: `from .util import foo` (this works everywhere)

## Problem 2: 

A Program wants to use the library and import `bar` from mylibrary:

1. If it references mylibrary as a project reference with same outdir: `from mylibrary.library import bar`
2. If it uses mylibrary from a Fable nuget: `from fable_modules.mylibrary import bar`
3. Relative: `ImportError: attempted relative import with no known parent package`

## Problem 3: 

You want to augment a python lib you are writing with some F# code

This is similar to fable_library where we have Fable.Library.fsproj that generates some extra python files such as map.py, mutable_set.py.

Now e.g `mutable_set` wants to import `FSharpRef` from `util.py`. What import should we generate?

Relative: `from .types import FSharpRef`
Absolute: `from fable_modules.fable_library.types import FSharpRef`. 

PS: The absolute is correct (when used), but will look wrong in the editor, so harder to look at the generated code:

<img width="941" alt="Screenshot 2021-11-05 at 08 05 06" src="https://user-images.githubusercontent.com/849479/140471611-36ea9670-d91a-4549-b956-37cda99f0905.png">

The problem is that we have no clue where this library will be mounted. For this example inside fable-library we could detect it, but for other libraries we have no way of generating a valid absolute import. E.g it could have been:

`from abc.def.types import FSharpRef`

## Summary

- We cannot generate valid absolute imports for libraries, we must use relative import
- We must use absolute imports for a program, we cannot use relative imports

Currently we detect this using `<OutputType>Exe</OutputType>` vs `<OutputType>Library</OutputType>`. You say that setting it to Library (or not setting) creates an invalid project? How? Here are some examples where I know it works:

- https://github.com/fable-compiler/Fable/blob/beyond/src/fable-library-py/fable_library/Fable.Library.fsproj
- https://github.com/fable-compiler/Fable.Python/blob/main/src/Fable.Python.fsproj
- https://github.com/dbrattli/AsyncRx/blob/main/src/FSharp.Control.AsyncRx.fsproj
- https://github.com/dbrattli/Feliz.ViewEngine/blob/master/src/Feliz.ViewEngine.fsproj

Perhaps one option could be to use absolute imports and have a command line argument to tell Fable its position in the library (e.g `fable_modules.fable_library`) and a command line argument to explicitly tell Fable to generate relative imports when we want it.

Or find a better way of detecting if we are a library or a program than using `OutputType`, but I currently haven't found any. 

What do you think?

**Comment by dbrattli** [comment]
PS: Just pushed -alpha-010 which should hopefully fix your failing tests (and more) @Zaid-Ajaj

**Comment by Zaid-Ajaj** [comment]
Hi @dbrattli I think I understand the problem above and from my experiments I believe it only occurs when the entry point project (exe) references projects that are in sibling directories. If the entry point project (exe) is in a parent directory then referencing projects seem to work fine. Though I don't know what happens when the referenced/shared projects themselves have cross-references 🤔 I will need to investigate this. 

The way I have worked around this problem for now in [Fable.SimpleJson.Python](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python) is by turning the project [into a library](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python/blob/master/build/Program.fs#L76) just before publishing to nuget 😓 it works but this shouldn't be the norm I hope when building libraries. 

Good news though is that I published this thing and now it is working like a charm 🚀 

![simple-json-python](https://user-images.githubusercontent.com/13316248/140916241-13c2372b-cc96-40f6-afbe-6423dd395b4e.png)

**Comment by dbrattli** [comment]
@Zaid-Ajaj Great work, and thanks for the feedback. I'll investigate this more in the weekend so hopefully we can find a better solution. I will also need to investigate:
- The dictionary assignment issue. I think it must be related to Dictionary having a comparer, since it's then a `MutableMap.fs` and not a Python `dict`. 
- Unions cannot be used as keys for dictionary. Do you btw have the error message here?

**Comment by dbrattli** [comment]
@Zaid-Ajaj I think everything discussed in this issue have been addressed now so closing this issue. Please re-open a new issue if you find anything  to not be working as expected.
